File: PetroManager-AI-complete/mobile/lib/services/sales_service.dart
import 'api_client.dart';

class SalesService {
  final ApiClient _client = ApiClient();

  Future<Map<String, dynamic>> createSale({
    required String shiftId,
    String? tankId,
    String? fuelType,
    double? quantity,
    required double unitPrice,
    String? description,
    String? paymentMethod,
  }) async {
    final result = await _client.post('/sales', {
      'shiftId': shiftId,
      if (tankId != null) 'tankId': tankId,
      if (fuelType != null) 'fuelType': fuelType,
      if (quantity != null) 'quantity': quantity,
      'unitPrice': unitPrice,
      if (description != null) 'description': description,
      if (paymentMethod != null) 'paymentMethod': paymentMethod,
    });
    return Map<String, dynamic>.from(result);
  }

  Future<List<dynamic>> findByShift(String shiftId) async {
    final result = await _client.get('/sales', query: {'shiftId': shiftId});
    return List<dynamic>.from(result);
  }
}

File: PetroManager-AI-complete/backend/src/modules/sales/sales.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { FuelType, ShiftStatus } from '@prisma/client';
import { PrismaService } from '../../common/prisma/prisma.service';

interface CreateSaleInput {
  shiftId: string;
  tankId?: string;
  fuelType?: FuelType;
  quantity?: number;
  unitPrice: number;
  description?: string;
  paymentMethod?: string;
}

@Injectable()
export class SalesService {
  constructor(private readonly prisma: PrismaService) {}

  async create(input: CreateSaleInput) {
    const shift = await this.prisma.shift.findUnique({ where: { id: input.shiftId } });
    if (!shift) throw new NotFoundException('شیفت پیدا نشد');
    if (shift.status !== ShiftStatus.OPEN) {
      throw new BadRequestException('فروش فقط برای شیفت باز قابل ثبت است');
    }

    const isFuelSale = !!input.tankId;
    if (isFuelSale && !input.quantity) {
      throw new BadRequestException('برای فروش سوخت، مقدار (quantity) الزامی است');
    }

    const totalAmount = isFuelSale
      ? Number(input.quantity) * input.unitPrice
      : input.unitPrice;

    return this.prisma.$transaction(async (tx) => {
      if (isFuelSale) {
        const tank = await tx.fuelTank.findUnique({ where: { id: input.tankId } });
        if (!tank) throw new NotFoundException('مخزن پیدا نشد');
        if (Number(tank.currentLevel) < Number(input.quantity)) {
          throw new BadRequestException('موجودی مخزن برای این فروش کافی نیست');
        }

        await tx.fuelTank.update({
          where: { id: input.tankId },
          data: { currentLevel: { decrement: input.quantity } },
        });
      }

      return tx.sale.create({
        data: {
          shiftId: input.shiftId,
          tankId: input.tankId,
          fuelType: input.fuelType,
          quantity: input.quantity,
          unitPrice: input.unitPrice,
          totalAmount,
          description: input.description,
          paymentMethod: input.paymentMethod,
        },
      });
    });
  }

  findByShift(shiftId: string) {
    return this.prisma.sale.findMany({
      where: { shiftId },
      orderBy: { soldAt: 'desc' },
    });
  }

  findByDateRange(from: Date, to: Date) {
    return this.prisma.sale.findMany({
      where: { soldAt: { gte: from, lte: to } },
      orderBy: { soldAt: 'desc' },
    });
  }
}
